server/odc-service/src/main/java/com/oceanbase/odc/service/flow/task/DBStructureComparisonFlowableTask.java

```diff
@@ -15,50 +15,129 @@
+    private volatile StructureComparisonContext context;
+    @Autowired
+    private StructureComparisonService service;
+
     @Override
-    protected DBStructureComparisonTaskResult start(Long taskId, TaskService taskService, DelegateExecution execution)
+    protected Void start(Long taskId, TaskService taskService, DelegateExecution execution)
             throws Exception {
+        TaskEntity taskEntity = taskService.detail(taskId);
+        if (taskEntity == null) {
+            throw new IllegalStateException("Can not find task entity by id " + taskId);
+        }
+        log.info("Structure comparison task starts, taskId={}, activityId={}", taskId,
+                execution.getCurrentActivityId());
+
+        DBStructureComparisonParameter parameters = FlowTaskUtil.getDBStructureComparisonParameter(execution);
+        Validate.notNull(parameters, "Structure comparison task parameters can not be null");
+        Long flowInstanceId = FlowTaskUtil.getFlowInstanceId(execution);
+
+        context = service.create(parameters, taskId, taskEntity.getCreatorId(), flowInstanceId);
+        taskService.start(taskId, context.getStatus());
         return null;
     }
 
+    @Override
+    protected boolean cancel(boolean mayInterruptIfRunning, Long taskId, TaskService taskService) {
+        if (context == null) {
+            throw new IllegalStateException("Context is null, structure comparison task may not be running");
+        }
+        boolean result = context.cancel(true);
+        log.info("structure comparison task has been cancelled, taskId={}, result={}", taskId, result);
+        taskService.cancel(taskId, context.getStatus());
+        return true;
+    }
+
+    @Override
+    public boolean isCancelled() {
+        if (context == null) {
+            return false;
+        }
+        return context.isCancelled();
+    }
+
     @Override
     protected boolean isSuccessful() {
-        return false;
+        if (context == null || !context.isDone()) {
+            return false;
+        }
+        try {
+            return context.get(1, TimeUnit.MILLISECONDS).getStatus() == TaskStatus.DONE;
+        } catch (Exception e) {
+            return false;
+        }
     }
 
     @Override
     protected boolean isFailure() {
-        return false;
+        if (context == null || !context.isDone()) {
+            return false;
+        }
+        return context.getStatus().getStatus() == TaskStatus.FAILED;
     }
 
     @Override
     protected void onTimeout(Long taskId, TaskService taskService) {
-
+        log.warn("Structure comparison task timeout, taskId={}", taskId);
     }
 
     @Override
-    protected void onProgressUpdate(Long taskId, TaskService taskService) {
+    protected void onFailure(Long taskId, TaskService taskService) {
+        log.warn("Structure comparison task failed, taskId={}", taskId);
+        DBStructureComparisonTaskResult result = context.getStatus();
```

Comment:
context may be null

Reply:
Indeed, but I see other FlowableTasks also have this problem.

Reply:
done